src/bsfs.c

```diff
@@ -1,34 +1,261 @@
+    // insert after existing node
+    bs_file_t* prev_link = table->buckets[bucket];
+    file->next = *prev_link;
+    *prev_link = file;
+  } else {
+    // insert at head
+    file->next = table->head;
+    table->head = file;
+    table->buckets[bucket] = &table->head;
+
+    if (file->next) {
+      // patch other bucket with new next pointer
+      size_t next_bucket =
+          oft_bucket_of(file->next->index, table->bucket_count);
+      table->buckets[next_bucket] = &file->next;
+    }
+  }
+}
+
+static int oft_remove(bs_oft_t* table, bs_file_t file) {
+  size_t bucket = oft_bucket_of(file->index, table->bucket_count);
+  if (!table->buckets[bucket]) {
+    return -EINVAL;
+  }
+
+  bs_file_t* prev_link = table->buckets[bucket];
+  for (; *prev_link != file &&
+         oft_matches_bucket(*prev_link, bucket, table->bucket_count);
+       prev_link = &(*prev_link)->next) {
+  }
+
+  if (*prev_link != file) {
+    return -EINVAL;
+  }
+
+  if (*table->buckets[bucket] == file &&
+      !oft_matches_bucket(file->next, bucket, table->bucket_count)) {
+    // `file` was the only entry in the bucket - empty it
+    table->buckets[bucket] = NULL;
+  }
+
+  *prev_link = file->next;
+
+  if (*prev_link) {
+    // patch other bucket with new next pointer
+    size_t next_bucket =
+        oft_bucket_of((*prev_link)->index, table->bucket_count);
+    table->buckets[next_bucket] = prev_link;
+  }
+
+  table->size--;
+
+  return 0;
+}
+
+static int oft_rehash(bs_oft_t* table, size_t bucket_count) {
+  int ret = oft_realloc_buckets(table, bucket_count);
+  if (ret < 0) {
+    return ret;
+  }
+
+  bs_file_t iter = table->head;
+  table->head = NULL;
+
+  while (iter) {
+    bs_file_t next = iter->next;
+    oft_do_insert(table, iter);
+    iter = next;
+  }
+
+  return 0;
+}
+
+static int oft_insert(bs_oft_t* table, bs_file_t file) {
+  if (table->size + 1 > table->bucket_count * OFT_MAX_LOAD_FACTOR) {
+    // note: still power of 2
+    int rehash_status = oft_rehash(table, 2 * table->bucket_count);
+    if (rehash_status < 0) {
+      return rehash_status;
+    }
+  }
+
+  oft_do_insert(table, file);
+  table->size++;
+
+  return 0;
+}
+
+int bs_oft_init(bs_oft_t* table) {
+  memset(table, 0, sizeof(bs_oft_t));
+  int ret = oft_realloc_buckets(table, OFT_INITIAL_BUCKET_COUNT);
+  if (ret < 0) {
+    return ret;
+  }
+  return -pthread_mutex_init(&table->lock, NULL);
+}
+
+void bs_oft_destroy(bs_oft_t* table) {
+  pthread_mutex_destroy(&table->lock);
+  free(table->buckets);
+
+  bs_file_t iter = table->head;
+  while (iter) {
+    bs_file_t next = iter->next;
+    destroy_open_file(iter);
+    iter = next;
+  }
+}
+
+int bs_oft_get(bs_oft_t* table, struct bs_open_level_impl* level,
+               bft_offset_t index, bs_file_t* file) {
+  int ret = -pthread_mutex_lock(&table->lock);
+  if (ret < 0) {
+    return ret;
+  }
+
+  bs_file_t existing_file = oft_find(table, index);
+  if (existing_file) {
+    *file = existing_file;
+    goto success;
```

Comment:
Why don't you use `else`?

Reply:
I don't know what I was thinking when I wrote that, but it's been cleaned up in cd13a9e.